lib/grpc/stream.ex

```diff
@@ -1,17 +1,433 @@
+
+  ## Returns
+    - A `GRPCStream` that emits the single gRPC message under demand.
+
+  ## Example
+
+      flow = GRPCStream.single(request, max_demand: 5)
+  """
+  @spec single(any(), Keyword.t()) :: t()
+  def single(input, opts \\ []) when is_struct(input),
+    do: build_grpc_stream([input], Keyword.merge(opts, unary: true))
+
+  @doc """
+  Converts a single gRPC request into a `Flow` pipeline while extracting and propagating context metadata from the `GRPC.Server.Stream`.
+
+  This is useful for unary gRPC calls that require metadata propagation within a Flow-based pipeline.
+
+  ## Parameters
+
+    - `input`: The enumerable stream of incoming gRPC messages.
+    - `materializer`: A `%GRPC.Server.Stream{}` struct representing the current gRPC stream context.
+    - `opts`: Optional keyword list for configuring the Flow or GenStage producer.
+
+  ## Behavior
+
+    - Automatically extracts gRPC metadata from the stream headers and injects it into the options as `:metadata`.
+
+  ## Returns
+
+    - A `GRPC.Stream` that emits the single gRPC message under demand.
+
+  ## Example
+
+      flow = GRPC.Stream.single_as_ctx(request, stream, max_demand: 10)
+
+  """
+  @spec single_as_ctx(any(), GRPC.Server.Stream.t(), Keyword.t()) :: t()
+  def single_as_ctx(input, %GRPC.Server.Stream{} = materializer, opts \\ [])
+      when is_struct(input) do
+    meta = get_headers(materializer) || %{}
+    single(input, Keyword.merge(opts, metadata: meta))
+  end
+
+  @doc """
+  Wraps an existing `Flow` into a `GRPC.Stream` struct.
+
+  > Note: `:join_with` is not supported in this mode and will raise an error if passed.
+
+  ## Parameters
+
+    - `flow`: A valid `Flow` pipeline.
+    - `opts`: Optional keyword options.
+
+  ## Returns
+
+    A `%GRPC.Stream{}` struct wrapping the flow.
+  """
+  @spec from_flow!(Flow.t(), Keyword.t()) :: t()
+  def from_flow!(%Flow{} = flow, opts \\ []) do
+    if Keyword.has_key?(opts, :join_with) do
+      raise ArgumentError, "join_with option is not supported for Flow input"
+    end
+
+    %__MODULE__{flow: flow, options: opts}
+  end
+
+  @doc """
+  Extracts the internal `Flow` pipeline from a `GRPC.Stream`.
+
+  Raises an `ArgumentError` if the `flow` has not been initialized.
+
+  ## Returns
+
+    The internal `Flow` struct.
+  """
+  @spec to_flow!(t()) :: Flow.t()
+  def to_flow!(%__MODULE__{flow: nil}) do
+    raise ArgumentError, "GRPC.Stream has no flow, initialize it with from/2"
+  end
+
+  def to_flow!(%__MODULE__{flow: flow}), do: flow
+
+  @doc """
+  Executes the flow and emits responses into the provided gRPC server stream.
+
+  ## Parameters
+
+    - `flow`: A `GRPC.Stream` struct containing the flow to be executed.
+    - `stream`: A `GRPC.Server.Stream` to which responses are sent.
+    - `opts`: Optional keyword list. Currently supported:
+    - `:dry_run` — If `true`, responses are not sent (used for testing or inspection).
```

Comment:
Missing indentation here.

In general, we can have a `## Options` section where we describe the options accepted in `opts`